File: go_algorithm/numbers/find_three_consecutive_integers_that_sum_to_a_given_number.go
package main

import "go_algorithm/test_util"

/*
	https://leetcode.com/problems/find-three-consecutive-integers-that-sum-to-a-given-number/

	Cho một số nguyên num, hãy trả về ba số nguyên liên tiếp (dưới dạng mảng đã sắp xếp) sao cho
	tổng của chúng bằng num.

	Nếu không thể biểu diễn num thành tổng của ba số nguyên liên tiếp, hãy trả về một mảng rỗng.

	Ý tưởng:

		Vì: n + (n + 1) + (n + 2) = num -> 3n + 3 = num -> n = num / 3 - 1

		=> Tức là num phải chia hết cho 3.
*/
func sumOfThree_TimeLimitExceeded(n int64) []int64 {
	for i := int64(0); i < n; i++ {
		if i + (i + 1) + (i + 2) == n {
			return []int64{i, i + 1, i + 2}
		}
	}
	return []int64{}
}

func sumOfThree(num int64) []int64 {
	if num % 3 != 0 {
		return []int64{}
	}
	n := (num - 3) / 3
	return []int64{n, n + 1, n + 2}
}

func main() {
	test_util.AssertSlice([]int64{}, sumOfThree(4))
	test_util.AssertSlice([]int64{10, 11, 12}, sumOfThree(33))
	test_util.AssertSlice([]int64{}, sumOfThree(455172614))
}

File: leetcode/src/main/java/numbers/FindThreeConsecutiveIntegersThatSumToAGivenNumber.java
package numbers;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

/*
	https://leetcode.com/problems/find-three-consecutive-integers-that-sum-to-a-given-number/

	Cho một số nguyên num, hãy trả về ba số nguyên liên tiếp (dưới dạng mảng đã sắp xếp) sao cho
	tổng của chúng bằng num.

	Nếu không thể biểu diễn num thành tổng của ba số nguyên liên tiếp, hãy trả về một mảng rỗng.

	Ý tưởng:

		Vì: n + (n + 1) + (n + 2) = num -> 3n + 3 = num -> n = num / 3 - 1

		=> Tức là num phải chia hết cho 3.
*/
public class FindThreeConsecutiveIntegersThatSumToAGivenNumber {

    public long[] sumOfThree(long num) {
        if (num % 3 != 0) return new long[]{};
        long n = num / 3 - 1;
        return new long[]{n, n + 1, n + 2};
    }

    @Test
    public void test() {
        assertArrayEquals(new long[]{}, sumOfThree(4));
        assertArrayEquals(new long[]{10, 11, 12}, sumOfThree(33));
        assertArrayEquals(new long[]{}, sumOfThree(455172614));
    }

}
